### PR TITLE
feat: cross chain model changes as per specs

### PR DESCRIPTION
**Pull request by hpsing**
* IsCrossChainIntent added
* SetEVMInstructions modified for cross chain
* encodeCrossChainCallData (as per specs `[2 bytes opType (0xFFFF)] + [2 bytes callDataLength] + [callData] + [32 bytes otherChainHash]`)
* test cases 

## Summary by CodeRabbit

- **New Features**
	- Introduced functions to extract source and destination chain IDs from intent objects.
	- Enhanced UserOperation structure to support cross-chain operations, including new methods for validating and encoding cross-chain intents.
	- Added a new structure for managing cross-chain data and related functionalities.
- **Bug Fixes**
	- Improved error handling for unsupported intent types and invalid chain IDs.
- **Tests**
	- Implemented comprehensive tests to validate chain ID extraction and cross-chain functionality in UserOperation, ensuring robust handling of various scenarios.

### CODE REVIEW

File: userops_ext.go [hunk 1]
```diff
@@ -304,6 +315,26 @@ func (op *UserOperation) GetIntentJSON() (string, error) {
...
+		return false, err
+	}
+
+	destChainID, err := ExtractDestinationChainID(intent)
+	if err != nil {
+		return false, err
+	}
+	// not equal chain ids.
+	return srcChainID.Cmp(destChainID) != 0, nil
+}
```
[the remark this answers is not in the document]
Comment (hpsing) [reply userops_ext.go]: not a senario as intent building validates chain id while creation

File: userops_ext_test.go [hunk 2]
```diff
@@ -145,47 +145,79 @@ func mockIntentJSON() string {
...
+	intentJSON := fmt.Sprintf(`
+    {
+    "fromAsset":{"address":"0x0A7199a96fdf0252E09F76545c1eF2be3692F46b","amount":%s,"chainId":%s},
+    "toAsset":{"address":"0x6B5f6558CB8B3C8Fec2DA0B1edA9b9d5C064ca47","amount":%s,"chainId":%s},
+    "extraData":{"partiallyFillable":false},
+    "status":"PROCESSING_STATUS_RECEIVED"}
+    `, fromB.String(), chainIDBuffer.String(), toB.String(), chainIDBuffer.String())
+
+	return intentJSON
+}
```
[the remark this answers is not in the document]
Comment (hpsing) [reply userops_ext_test.go]: this is intentional to diffrentiate and extend the case for future

**Review by adelowo — APPROVED** [review]

File: userops_ext.go [hunk 3]
```diff
@@ -400,63 +432,124 @@ func (op *UserOperation) GetSignatureValue() []byte {
...
 
+	// Validate the UserOperation's current state
 	status, err := op.Validate()
 	if err != nil {
-		return err
+		return fmt.Errorf("invalid UserOperation state: %w", err)
+	}
+
+	// Extract any existing intent
+	intentJSON, hasIntent := op.extractIntentJSON()
```
Comment (blewater) [inline userops_ext.go]: extractIntentJSON needs to be enhanced to extract the Intent from the x-chain calldata field, i.e.,
- **CallData Format:**
    - `[2 bytes opType (0xFFFF)]`
    - `[2 bytes length of intent JSON]`
    - `[Intent JSON]`
    - `[1 byte hash list length (N)]`
    - `[`**Hash List Entry**`]`

File: userops_ext_test.go [hunk 4]
```diff
@@ -198,6 +231,15 @@ func mockUserOperationWithIntentInCallData() *UserOperation {
 	return userOp
 }
 
+func mockUserOperationWithCrossChainIntentInCallData() *UserOperation {
+	userOp := new(UserOperation)
+	intentJSON := mockCrossChainIntentJSON()
+
+	userOp.CallData = []byte(intentJSON)
```
Comment (blewater) [inline userops_ext_test.go:246]: - **CallData Format:**
    - [2 bytes opType (0xFFFF)]
    - [2 bytes length of intent JSON]
    - `[Intent JSON]`
    - [1 byte hash list length (N)]
    - [**Hash List Entry**]

**Review by blewater — CHANGES_REQUESTED** [review]
Sorry, I didn't explain the x-chain format adequately:
Unsolved x-chain format:
```
- **CallData Field Format:**
    - [2 bytes opType (0xFFFF)]
    - [2 bytes length of intent JSON]
    - `[Intent JSON]`
    - [1-byte hash list length (N)]
    - [**Hash List Entry**]
```
When Solved:
Same format as signature suffix (past 65 bytes)

File: userops_ext.go [hunk 5]
```diff
@@ -432,93 +462,59 @@ func (op *UserOperation) GetSignatureValue() []byte {
...
 	if hasIntent {
 		if !op.HasSignature() {
+			// Need a signed userOp to append the Intent JSON to the signature value.
 			return ErrNoSignatureValue
 		}
+
 		op.Signature = append(op.GetSignatureValue(), []byte(intentJSON)...)
-		op.CallData = nil
+		// Clear the Intent JSON from CallData as it's now moved to Signature.
```
Comment (blewater) [inline userops_ext.go]: The comment should move below outside the bracket before the final assignment

File: userops_ext.go [hunk 6]
```diff
@@ -432,93 +462,59 @@ func (op *UserOperation) GetSignatureValue() []byte {
...
-		} else {
-			op.CallData = callDataValue
-		}
+		op.CallData = callDataValue
 		return nil
 	}
 
-	// Handle unsolved operations
+	// Unsolved operation, move the Intent JSON to the Signature field if it exists.
+	intentJSON, hasIntent := op.extractIntentJSON()
```
Comment (blewater) [inline userops_ext.go:473]: For x-chain ops, the entire validated calldata value appends the signature:
- `[2 bytes opType (0xFFFF)]`
- `[2 bytes length of intent JSON]`
- `[Intent JSON]`
- `[1 byte hash list length (N)]`
- `[`**Hash List Entry**`]`

File: userops_ext.go [hunk 7]
```diff
@@ -183,11 +183,41 @@ func no0xPrefix(value []byte) bool {
...
+		binary.BigEndian.Uint16(dataBytes[:OpTypeLength]) == CrossChainMarker {
+
+		intentLengthBytes := dataBytes[2:4]
+		intentLength := int(binary.BigEndian.Uint16(intentLengthBytes))
+
+		if len(dataBytes) >= 4+intentLength {
+			intentJSON := string(dataBytes[4 : 4+intentLength])
+
+			if len(dataBytes) > 4+intentLength {
+				return intentJSON, true
```
Comment (blewater) [inline userops_ext.go]: Add proto unmarshal validation for the bytes like for uni-chain intent bytes. See `ExtractJSONFromField()` comment.

File: userops_ext_test.go [hunk 8]
```diff
@@ -1153,29 +1156,12 @@ func TestUserOperation_IsCrossChainIntent(t *testing.T) {
...
-}
-
 // TestValidateUserOperation_CrossChain test Validate for cross chain intent.
 func TestValidateUserOperation_CrossChain(t *testing.T) {
-	t.Run("Validate cross-chain UserOperation", func(t *testing.T) {
-		uop := mockUserOperationWithCrossChainIntentInCallData()
-		status, err := uop.Validate()
-		require.NoError(t, err)
-		require.Equal(t, UnsolvedUserOp, status)
-	})
```
Comment (blewater) [inline userops_ext_test.go]: We should cover cross-chain op validation with a comprehensive test suite.

File: userops_ext.go [hunk 9]
```diff
@@ -432,93 +462,59 @@ func (op *UserOperation) GetSignatureValue() []byte {
...
+		// `Decode` allows using the source as the destination
 		var err error
 		callDataValue, err = hexutil.Decode(string(callDataValue))
 		if err != nil {
 			return fmt.Errorf("invalid hex encoding of calldata: %w", err)
 		}
 	}
 
-	// Validate the UserOperation's current state
 	status, err := op.Validate()
```
Comment (blewater) [inline userops_ext.go]: Validate() should be enhanced for cross-chain ops to validate intent JSON and struct
- `[2 bytes opType (0xFFFF)]`
- `[2 bytes length of intent JSON]`
- `[Intent JSON]`
- `[1 byte hash list length (N)]`
- `[`**Hash List Entry**`]`

File: userops_ext.go [hunk 10]
```diff
@@ -171,11 +183,41 @@ func no0xPrefix(value []byte) bool {
...
+			if len(dataBytes) > 4+intentLength {
+				return intentJSON, true
+			}
+		}
+
+		return "", false
+	}
+
 	// Try to extract Intent JSON from CallData field
 	if intentJSON, ok := ExtractJSONFromField(string(op.CallData)); ok {
```
Comment (blewater) [inline userops_ext.go]: Enhance `ExtractJSONFromField` to support cross-chain ops
- `[2 bytes opType (0xFFFF)]`
- `[2 bytes length of intent JSON]`
- `[Intent JSON]`
- `[1 byte hash list length (N)]`
- `[`**Hash List Entry**`]`

**Review by blewater — CHANGES_REQUESTED** [review]
Getting there but requires work still.

File: userops_ext.go [hunk 11]
```diff
@@ -171,11 +281,33 @@ func no0xPrefix(value []byte) bool {
...
+// extractIntentJSON attempts to extract the Intent JSON from either the CallData
+// or Signature field of a UserOperation.
 func (op *UserOperation) extractIntentJSON() (string, bool) {
+	// Try to check if it is a cross chain calldata
+	if isCrossChain, _ := op.isCrossChainOperation(); isCrossChain {
+		dataBytes := op.CallData
+		intentLength := int(binary.BigEndian.Uint16(dataBytes[OpTypeLength : OpTypeLength+CallDataLengthSize]))
+		if len(dataBytes) >= OpTypeLength+CallDataLengthSize+intentLength {
+			intentJSON := string(dataBytes[OpTypeLength+CallDataLengthSize : OpTypeLength+CallDataLengthSize+intentLength])
+			return intentJSON, true
```
Comment (blewater) [inline userops_ext.go:204]: This logic branch does not validate the Intent JSON and presumes an unsolved Operation.

File: userops_ext.go [hunk 12]
```diff
@@ -161,6 +193,84 @@
...
+
+	return !isAllZeros
+}
+
+// isCrossChainOperation checks if the UserOperation is a cross-chain operation
+func (op *UserOperation) isCrossChainOperation() (bool, error) {
+	if len(op.CallData) < OpTypeLength {
+		return false, nil
+	}
+	return binary.BigEndian.Uint16(op.CallData[:OpTypeLength]) == CrossChainMarker, nil
```
[the remark this answers is not in the document]
Comment (blewater) [reply userops_ext.go]: It works only for unsolved ops; the x-payload would be appended to the signature payload in solved ops.

File: userops_ext.go [hunk 13]
```diff
@@ -211,6 +343,61 @@ func ExtractJSONFromField(fieldData string) (string, bool) {
...
+	if err != nil {
+		return err
+	}
+
+	srcChainID, err := ExtractSourceChainID(intent)
+	if err != nil {
+		return err
+	}
+
+	destChainID, err := ExtractDestinationChainID(intent)
```
Comment (blewater) [inline userops_ext.go]: Return an error if the dest chain id == source chain id

File: userops_ext.go [hunk 14]
```diff
@@ -161,6 +193,84 @@
...
+			return UnknownUserOp, ErrInvalidHashListEntry
+		}
+	}
+
+	if op.HasSignature() {
+		return SolvedUserOp, nil
+	}
+
+	return UnsolvedUserOp, nil
+}
```
[the remark this answers is not in the document]
Comment (blewater) [reply userops_ext.go:245]: Look like good suggestions

File: userops_ext.go [hunk 15]
```diff
@@ -211,6 +343,61 @@ func ExtractJSONFromField(fieldData string) (string, bool) {
...
+	destChainID, err := ExtractDestinationChainID(intent)
+	if err != nil {
+		return err
+	}
+
+	crossChainData := make([]byte, OpTypeLength+CallDataLengthSize+len(intentJSON)+HashListLengthSize)
+	binary.BigEndian.PutUint16(crossChainData[:OpTypeLength], CrossChainMarker)
+	binary.BigEndian.PutUint16(crossChainData[OpTypeLength:OpTypeLength+CallDataLengthSize], uint16(len(intentJSON)))
+	copy(crossChainData[OpTypeLength+CallDataLengthSize:], []byte(intentJSON))
+	crossChainData[OpTypeLength+CallDataLengthSize+len(intentJSON)] = 2 // Hash list length
```
Comment (blewater) [inline userops_ext.go]: These lines should have unit test coverage.

**Review by blewater — CHANGES_REQUESTED** [review]
https://github.com/blndgs/model/blob/93f98c4f4b39bbabfd95bf2a2485468c18722f1e/userops_ext.go#L646
For cross-chain userOps append the entire x-payload or the signature won't validate.

File: userops_ext.go [hunk 16]
```diff
@@ -161,6 +193,84 @@ func (op *UserOperation) Validate() (UserOpSolvedStatus, error) {
...
+
+	return !isAllZeros
+}
+
+// isCrossChainOperation checks if the UserOperation is a cross-chain operation
+func (op *UserOperation) isCrossChainOperation() (bool, error) {
+	if len(op.CallData) < OpTypeLength {
+		return false, nil
+	}
+	return binary.BigEndian.Uint16(op.CallData[:OpTypeLength]) == CrossChainMarker, nil
```
Comment (blewater) [inline userops_ext.go]: It works only for unsolved ops; the x-payload would be appended to the signature payload in solved ops.

File: userops_ext.go [hunk 17]
```diff
@@ -161,6 +192,104 @@ func (op *UserOperation) Validate() (UserOpSolvedStatus, error) {
...
+	if len(op.CallData) >= OpTypeLength &&
+		binary.BigEndian.Uint16(op.CallData[:OpTypeLength]) == CrossChainMarker {
+		return true
+	}
+
+	// solved operation has it's cross-chain data in the signature
+	idx := op.GetSignatureEndIdx()
+	if idx > 0 && idx < len(op.Signature) {
+		xData := op.Signature[idx:]
+		if len(xData) >= OpTypeLength &&
```
Comment (blewater) [inline userops_ext.go]: if len(xData) == OpTypeLength, it's not a crosschain operation. Add additional length checks to enhance detection according to the spec, e.g.., length > 2 + json length + 32 + 2

File: userops_ext.go [hunk 18]
```diff
@@ -161,6 +192,104 @@ func (op *UserOperation) Validate() (UserOpSolvedStatus, error) {
...
+			break
+		}
+	}
+
+	return !isAllZeros
+}
+
+// isCrossChainOperation checks if the UserOperation is a cross-chain operation
+func (op *UserOperation) isCrossChainOperation() bool {
+	if len(op.CallData) >= OpTypeLength &&
```
Comment (blewater) [inline userops_ext.go]: See next comment about equality comparison.

File: userops_ext.go [hunk 19]
```diff
@@ -595,3 +844,64 @@ func (op *UserOperation) String() string {
 		formatBytes(op.Signature),
 	)
 }
+
+// encodeCrossChainCallData encodes the call data in the cross-chain format
+// [2 bytes opType (0xFFFF)] + [2 bytes callDataLength] + [callData] + [1 byte hashListLength] + [Hash List Entries]
```
Comment (blewater) [inline userops_ext.go]: These came from an older documentation version, please update the names in the code.
`callDataLength` is IntentJSON length
`callData` is Intent JSON

File: userops_ext.go [hunk 20]
```diff
@@ -595,3 +844,64 @@ func (op *UserOperation) String() string {
...
+		return nil, fmt.Errorf("failed to get intent: %w", err)
+	}
+
+	// Extract source and destination chain IDs
+	sourceChainID, err := ExtractSourceChainID(intent)
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract source chain ID: %w", err)
+	}
+
+	destChainID, err := ExtractDestinationChainID(intent)
```
Comment (blewater) [inline userops_ext.go]: Validate srcChainID != destChainID etc. You may centralize the validation rule.

File: userops_ext.go [hunk 21]
```diff
@@ -439,16 +673,24 @@ func (op *UserOperation) SetEVMInstructions(callDataValue []byte) error {
...
+	if !op.HasSignature() {
+		return ErrNoSignatureValue
+	}
+
+	// Check if it's a cross-chain intent
+	switch op.isCrossChainOperation() {
+	case true:
+		// For cross-chain intents, we need to include the whole cross-chain spec in the signature
+		// safe to righfully assert the callData contains the cross-chain spec for unsolved ops
+		crossChainData := op.CallData
```
Comment (blewater) [inline userops_ext.go]: If it's a solved operation, it should not append anything to the signature.

File: userops_ext.go [hunk 22]
```diff
@@ -595,3 +844,64 @@ func (op *UserOperation) String() string {
 		formatBytes(op.Signature),
 	)
 }
+
+// encodeCrossChainCallData encodes the call data in the cross-chain format
+// [2 bytes opType (0xFFFF)] + [2 bytes callDataLength] + [callData] + [1 byte hashListLength] + [Hash List Entries]
+func (op *UserOperation) encodeCrossChainCallData(callData []byte) ([]byte, error) {
```
Comment (blewater) [inline userops_ext.go]: This function (encodeCrossChainCallData) is done only by the SDK. Please move it to the test file if you need it for mocking.

File: userops_ext.go [hunk 23]
```diff
@@ -595,3 +844,64 @@ func (op *UserOperation) String() string {
...
+	// Source chain entry
+	binary.BigEndian.PutUint16(crossChainCallData[hashListStart:], CrossChainMarker) // Placeholder
+	sourceHash := common.BigToHash(sourceChainID)
+	copy(crossChainCallData[hashListStart+PlaceholderSize:], sourceHash[:])
+
+	// Destination chain entry
+	destEntryStart := hashListStart + HashListEntrySize
+	binary.BigEndian.PutUint16(crossChainCallData[destEntryStart:], CrossChainMarker) // Placeholder
+	destHash := common.BigToHash(destChainID)
+	copy(crossChainCallData[destEntryStart+PlaceholderSize:], destHash[:])
```
Comment (blewater) [inline userops_ext.go]: This part is incorrect--see spec.

**Review by blewater — CHANGES_REQUESTED** [review]
Thx for the enhancements... a few more

File: userops_ext.go (hunk 22 above)
Comment (blewater) [reply userops_ext.go]: oh the CLI will do this besides the SDK...

File: userops_ext.go [hunk 24]
```diff
@@ -595,3 +911,58 @@ func (op *UserOperation) String() string {
 		formatBytes(op.Signature),
 	)
 }
+
+// encodeCrossChainCallData encodes the call data in the cross-chain format
+func (op *UserOperation) encodeCrossChainCallData(intentJSONBytes []byte) ([]byte, error) {
```
Comment (blewater) [inline userops_ext.go]: it's only used by the tests. Why not move it to the test file as a mocking function?

File: userops_ext.go [hunk 25]
```diff
@@ -595,3 +911,58 @@ func (op *UserOperation) String() string {
...
+	offset++
+
+	binary.BigEndian.PutUint16(crossChainCallData[offset:], CrossChainMarker)
+	offset += OpTypeLength
+	copy(crossChainCallData[offset:], common.BigToHash(sourceChainID).Bytes())
+	offset += HashLength
+
+	binary.BigEndian.PutUint16(crossChainCallData[offset:], CrossChainMarker)
+	offset += OpTypeLength
+	copy(crossChainCallData[offset:], common.BigToHash(destChainID).Bytes())
```
Comment (blewater) [inline userops_ext.go]: This bytes section is wrong, and i can work on it.

**Review by blewater — CHANGES_REQUESTED** [review]
Looks great, but encodeCrossChainCallData needs work. I can finish it.

**Review by blewater — APPROVED** [review]

File: userops_ext.go [hunk 26]
```diff
@@ -320,51 +353,40 @@
...
+	// since we rely  on the isCrossChain variable to determine if to set
+	// the cross chain item
+	//
+	// err could be like no intent json or not a cross chain operation
+	// so safe to ignore here.
+	isCrossChain, _ := op.IsCrossChainIntent()
+	if isCrossChain {
+		return op.setCrossChainIntent(intentJSON)
+	}
+
```
[the remark this answers is not in the document]
Comment (blewater) [reply userops_ext.go:375]: check the inline comment, safe to ignore

File: userops_ext_xchain.go [hunk 27]
```diff
@@ -0,0 +1,398 @@
...
+		return false
+	}
+
+	hashListLength := len(crossChainData.HashList)
+	if hashListLength < minHashListLength || (maxHashListLength >= 0 && hashListLength > maxHashListLength) {
+		return false
+	}
+
+	return true
+}
```
[the remark this answers is not in the document]
Comment (blewater) [reply userops_ext_xchain.go]: has additional rule checks not in the Parse function.

**Pull request merged**